### PR TITLE
SWC-6800, SWC-6938 - Refactor to create generic AclEditor

### PR DESCRIPTION
**Pull request by nickgros**
Refactor work to support SWC-6800
- Extract new component AclEditor & useUpdateAcl from AccessRequirementAclEditor component
- Update text to add a user to the ACL (SWC-6938)
- Add animation to list of users & teams to draw the user's attention to the change

### CODE REVIEW

File: packages/synapse-react-client/src/components/AccessRequirementAclEditor/AccessRequirementAclEditor.test.tsx [hunk 1]
```diff
@@ -86,62 +91,6 @@ async function setUp(
   return { ref, component, itemRows, user }
 }
 
-const confirmItem = (
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/AccessRequirementAclEditor/AccessRequirementAclEditor.test.tsx:89]: Moved these utilities to new file AclEditorTestUtils so all ACL editor tests can share them. Made small tweaks to make them more generic & reusable.

File: packages/synapse-react-client/src/components/AccessRequirementAclEditor/AccessRequirementAclEditor.test.tsx [hunk 2]
```diff
@@ -389,9 +342,15 @@ describe('AccessRequirementAclEditor', () => {
...
 
     const alert = screen.getByRole('alert')
-    expect(alert).toHaveTextContent(REVIEWER_ALREADY_ADDED_ERROR_MESSAGE)
+    expect(alert).toHaveTextContent(PRINCIPAL_ALREADY_ADDED_ERROR_MESSAGE)
+
+    // Verify the error goes away when a successful change is made
+    const newUserRow = await addUserToAcl(user, MOCK_USER_NAME_2)
+    confirmItem(newUserRow, MOCK_USER_NAME_2, 'Can Review')
+
+    expect(alert).not.toBeInTheDocument()
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/AccessRequirementAclEditor/AccessRequirementAclEditor.test.tsx:354]: Add assertion that error disappears

File: packages/synapse-react-client/src/components/AccessRequirementAclEditor/AccessRequirementAclEditor.tsx [hunk 3]
```diff
@@ -65,12 +63,23 @@ export const AccessRequirementAclEditor = React.forwardRef(
...
+    const {
+      resourceAccessList,
+      setResourceAccessList,
+      addResourceAccessItem,
+      updateResourceAccessItem,
+      removeResourceAccessItem,
+    } = useUpdateAcl({
+      onChange: () => setError(null),
+      onError: setError,
+    })
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/AccessRequirementAclEditor/AccessRequirementAclEditor.tsx:75]: Storing ACL in state & making changes to it is captured in this hook

File: packages/synapse-react-client/src/components/AccessRequirementAclEditor/AccessRequirementAclEditor.tsx [hunk 4]
```diff
@@ -196,67 +160,15 @@ export const AccessRequirementAclEditor = React.forwardRef(
...
-              </Typography>
-              <UserSearchBoxV2
-                inputId="reviewer-search"
-                placeholder="Username, name (first and last) or team name."
-                onChange={addResourceAccessItem}
-              />
-            </Box>
-          </Box>
-        </Box>
+        <AclEditor
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/AccessRequirementAclEditor/AccessRequirementAclEditor.tsx:163]: Use new AclEditor component

File: packages/synapse-react-client/src/components/AclEditor/AclEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    addResourceAccessItem,
+    updateResourceAccessItem,
+    removeResourceAccessItem,
+  } = props
+  if (isLoading) {
+    return (
+      <Box display={'flex'} justifyContent={'center'} height={'150px'}>
+        <SynapseSpinner size={50} />
+      </Box>
+    )
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/AclEditor/AclEditor.tsx:42]: Tweaked the loading UI to only show a spinner if loading. There was a skeleton in the old code but it was one rectangle & didn't block the rest of the UI, so this spinner is an improvement. May revisit with a more fleshed out skeleton later.

File: packages/synapse-react-client/src/components/AclEditor/AclEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+          Users and Teams with Permissions
+        </Typography>
+
+        {resourceAccessList.length === 0 ? (
+          <Typography variant="body1Italic">{emptyText}</Typography>
+        ) : (
+          <TransitionGroup>
+            {resourceAccessList.map(resourceAccess => {
+              return (
+                <Collapse key={resourceAccess.principalId}>
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/AclEditor/AclEditor.tsx:58]: Add `TransitionGroup`/`Collapse` to animate addition/removal of items. This feels much more natural in cases like this when the UI to add a row is disconnected from the rest of the form.

[Screencast from 07-11-2024 05:37:47 PM.webm](https://github.com/user-attachments/assets/56c51112-84e3-4cb6-9eb6-7cb5958a2cca)

File: packages/synapse-react-client/src/components/AclEditor/ResourceAccessItem.tsx [hunk 7]
```diff
@@ -8,22 +8,18 @@ import { IconSvgButton } from '../IconSvgButton'
 
 export const REMOVE_BUTTON_LABEL = 'Remove from AR Permissions'
 
-const availablePermissionLevels: PermissionLevel[] = [
-  'CAN_REVIEW_SUBMISSIONS',
-  'IS_EXEMPTION_ELIGIBLE',
-  'CAN_REVIEW_SUBMISSIONS_AND_IS_EXEMPTION_ELIGIBLE',
-]
-
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/AclEditor/ResourceAccessItem.tsx:16]: These are Access Requirement-specific permission levels, so move them to the AR ACL editor file. `availablePermissionLevels` is now passed as a prop so each editor can be configured for the resource it supports.

File: packages/synapse-react-client/src/components/CreateOrUpdateAccessRequirementWizard/CreateOrUpdateAccessRequirementWizard.integration.test.tsx [hunk 8]
```diff
@@ -239,7 +239,9 @@ describe('CreateOrUpdateAccessRequirementWizard', () => {
     })
     expect(removeBtns).toHaveLength(nPrincipalsOnACL)
     await user.click(removeBtns[nPrincipalsOnACL - 1])
-    expect(userOnAcl).not.toBeInTheDocument()
+    await waitFor(() => {
+      expect(userOnAcl).not.toBeInTheDocument()
+    })
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/CreateOrUpdateAccessRequirementWizard/CreateOrUpdateAccessRequirementWizard.integration.test.tsx:244]: The new animation requires us to wait a little longer for the row to disappear

File: packages/synapse-react-client/src/components/AclEditor/AclEditor.tsx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    resourceAccessList,
+    availablePermissionLevels,
+    emptyText,
+    addResourceAccessItem,
+    updateResourceAccessItem,
+    removeResourceAccessItem,
+  } = props
+  if (isLoading) {
+    return (
+      <Box display={'flex'} justifyContent={'center'} height={'150px'}>
```
Comment (jay-hodgson) [inline packages/synapse-react-client/src/components/AclEditor/AclEditor.tsx:39]: One stylistic question - I've been using sx styles to change these, which I've found really convenient for rapid iterative development.  Is there a reason to use the specific component properties (like `display`), or is this just a preference?

**Review by jay-hodgson — APPROVED** [review]

File: packages/synapse-react-client/src/components/AclEditor/AclEditor.tsx (hunk 9 above)
Comment (nickgros) [reply packages/synapse-react-client/src/components/AclEditor/AclEditor.tsx:39]: I tend to use the props if they are there, [which is true for Box, Stack, Typography, and Grid](https://mui.com/system/getting-started/usage/#api-tradeoff)

For cases where we would want to share the same set of styles across multiple `Box` instances, we'd probably be better of using a shared `sx` object or styled component. For one-offs, I don't really have a preference. The props feel more natural to me.

**Pull request merged**